File: UVA/AOAPC-BAC/Ch3/Example/UVA-401_Palindromes.cpp
/*
    tags: string, palindrome, 紫書-Ch3, CPE-140325
*/
#include <bits/stdc++.h>
using namespace std;
#define endl '\n'

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
    int n, idx;
    bool flag1, flag2;
    string s, mp = "A   3  HIL JM O   2TUVWXY51SE Z  8 ";
    while (getline(cin, s)) {
        n = s.size();
        flag1 = (s == string(s.rbegin(), s.rend()));
        flag2 = true;
        for (int i = 0; i < n/2+1; ++i) {
            if (isupper(s[i])) idx = s[i] - 'A';
            else if (isdigit(s[i])) idx = s[i] - '0' + 25;
            else idx = -1;
            if (idx == -1 || s[n-1-i] != mp[idx]) {
                flag2 = false;
                break;
            }
        }
        cout << s << " -- is ";
        if (!flag1 && !flag2) cout << "not a palindrome." << endl;
        else if (flag1 && !flag2) cout << "a regular palindrome." << endl;
        else if (!flag1 && flag2) cout << "a mirrored string." << endl;
        else cout << "a mirrored palindrome." << endl;
        cout << endl;
    }
    return 0;
}

File: UVA/AOAPC-BAC/Ch3/Example/UVA-401_Palindromes.py
"""
    tags: string, palindrome, 紫書-Ch3, CPE-140325
"""
mp = "A   3  HIL JM O   2TUVWXY51SE Z  8 "

while True:
    try:
        s = input()
    except EOFError:
        break
    n = len(s)
    flag1 = (s == s[::-1]) # is a palindrome
    flag2 = True
    for i in range(n//2+1):
        if s[i].isupper():
            idx = ord(s[i]) - ord('A')
        elif s[i].isdigit():
            idx = ord(s[i]) - ord('0') + 25
        else:
            idx = -1
        if idx == -1 or s[n-1-i] != mp[idx]:
            flag2 = False
            break
    print(s, end=" -- is ")
    if not flag1 and not flag2:
        print("not a palindrome.")
    elif flag1 and not flag2:
        print("a regular palindrome.")
    elif not flag1 and flag2:
        print("a mirrored string.")
    else:
        print("a mirrored palindrome.")
    print()